server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -356,92 +334,16 @@ private <C extends Collector, T> T search(Weight weight, CollectorManager<C, T>
                     throw new IllegalStateException("CollectorManager does not always produce collectors with the same score mode");
                 }
             }
-            final List<RunnableFuture<C>> listTasks = new ArrayList<>();
+            final List<Callable<C>> listTasks = new ArrayList<>();
             for (int i = 0; i < leafSlices.length; ++i) {
                 final LeafReaderContext[] leaves = leafSlices[i].leaves;
                 final C collector = collectors.get(i);
-                AtomicInteger state = new AtomicInteger(0);
-                RunnableFuture<C> task = new FutureTask<>(() -> {
-                    if (state.compareAndSet(0, 1)) {
-                        // A slice throws exception or times out: cancel all the tasks, to prevent slices that haven't started yet from
-                        // starting and performing needless computation.
-                        // TODO we will also want to cancel tasks that have already started, reusing the timeout mechanism
-                        try {
-                            search(Arrays.asList(leaves), weight, collector);
-                            if (timeExceeded) {
-                                for (Future<?> future : listTasks) {
-                                    FutureUtils.cancel(future);
```

Comment:
we used to have cancellation on timeout as well. This is not currently implemented in Lucene (we cancel only on exceptions), and looking further at implementing that, I realized that it's perhaps not needed, because if one task times out, it is expected that others will do too, as the timeout impl is shared across the different tasks. That is even better, as it will make running tasks stop too, which is something that tasks cancellation does not support yet.